app/src/main/java/com/asfoundation/wallet/verification/intro/VerificationIntroPresenter.kt

```diff
@@ -0,0 +1,195 @@
+package com.asfoundation.wallet.verification.intro
+
+import android.os.Bundle
+import com.appcoins.wallet.billing.adyen.VerificationPaymentModel
+import com.appcoins.wallet.billing.common.response.TransactionStatus
+import com.appcoins.wallet.billing.util.Error
+import com.asfoundation.wallet.billing.adyen.AdyenErrorCodeMapper
+import com.asfoundation.wallet.logging.Logger
+import io.reactivex.Completable
+import io.reactivex.Scheduler
+import io.reactivex.disposables.CompositeDisposable
+import java.util.concurrent.TimeUnit
+
+class VerificationIntroPresenter(private val view: VerificationIntroView,
+                                 private val disposable: CompositeDisposable,
+                                 private val navigator: VerificationIntroNavigator,
+                                 private val logger: Logger,
+                                 private val viewScheduler: Scheduler,
+                                 private val ioScheduler: Scheduler,
+                                 private val interactor: VerificationIntroInteractor,
+                                 private val adyenErrorCodeMapper: AdyenErrorCodeMapper,
+                                 private val data: VerificationIntroData) {
+
+  companion object {
+
+    private val TAG = VerificationIntroPresenter::class.java.name
+  }
+
+  fun present(savedInstanceState: Bundle?) {
+    loadModel(savedInstanceState)
+    handleCancelClicks()
+    handleForgetCardClick()
+    handleRetryClick(savedInstanceState)
+    handleTryAgainClicks()
+    handleSupportClicks()
+  }
+
+  private fun loadModel(savedInstanceState: Bundle?) {
+    disposable.add(
+        interactor.loadVerificationIntroModel()
+            .subscribeOn(ioScheduler)
+            .observeOn(viewScheduler)
+            .doOnSuccess {
+              view.finishCardConfiguration(it.paymentInfoModel.paymentMethodInfo!!,
+                  it.paymentInfoModel.isStored, false, savedInstanceState)
+              view.updateUi(it)
+              view.hideLoading()
+              view.unlockRotation()
+              handleSubmitClicks(it.verificationInfoModel)
+            }
+            .doOnSubscribe {
+              view.lockRotation()
+              view.showLoading()
+            }
+            .subscribe({}, {
+              logger.log(TAG, it)
+              view.showGenericError()
+            })
+    )
+  }
+
+  private fun handleCancelClicks() {
+    disposable.add(
+        view.getCancelClicks()
+            .doOnNext { view.cancel() }
+            .subscribe({}, { it.printStackTrace() })
+    )
+  }
+
+  private fun handleRetryClick(savedInstanceState: Bundle?) {
+    disposable.add(view.retryClick()
+        .observeOn(viewScheduler)
+        .doOnNext { view.showLoading() }
+        .delay(1, TimeUnit.SECONDS)
+        .doOnNext { loadModel(savedInstanceState) }
+        .subscribe({}, { it.printStackTrace() }))
```

Comment:
Add observeOn(viewScheduler) as delay runs on computation and may crash

Reply:
This isn't really an issue because loadModel handles this internally. Although I don't think this is the best way to do this (having a subscription inside a subscription).